### ISSUE TITLE
pkg/dds/test/dap/integration/debug_attach_test fails on pkg-win-release

### ISSUE THREAD

**Issue by alexmarkov** [dart-lang/sdk#48274]
There are new test failures on [[vm/compiler] Avoid speculative conversion in ffi Pointer.asTypedList](https://github.com/dart-lang/sdk/commit/e4942db83e2ad147de8710c0c96629f676fc91e3).

The tests
```
pkg/dds/test/dap/integration/debug_attach_test RuntimeError (expected Pass)
```
are failing on configurations
```
unittest-asserts-release-win
```

Log 1:
```
--- Command "vm" (took 17.000585s):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set CHROME_PATH=C:\b\s\w\ir\cache\builder\sdk\third_party\browsers\chrome\Chrome\Application\chrome.exe & out\ReleaseX64\dart-sdk\bin\dart.exe --enable_asserts -Dtest_runner.configuration=unittest-asserts-release-win --ignore-unrecognized-flags --packages=C:\b\s\w\ir\cache\builder\sdk\.packages C:\b\s\w\ir\cache\builder\sdk\pkg\dds\test\dap\integration\debug_attach_test.dart

exit code:
255

stdout:
00:00 [32m+0[0m: debug mode can attach to a simple script using vmServiceUri[0m

00:03 [32m+1[0m: debug mode can attach to a simple script using vmServiceInfoFile[0m

00:06 [32m+1[0m[31m -1[0m: debug mode can attach to a simple script using vmServiceInfoFile [1m[31m[E][0m[0m

  FileSystemException: Deletion failed, path = 'C:\b\s\w\ir\x\t\dart-sdk-dap-test8f36c1be' (OS Error: The process cannot access the file because it is being used by another process.

  , errno = 32)

  dart:io/directory_impl.dart 201:7                      _Directory._deleteSync
  dart:io/file_system_entity.dart 420:7                  FileSystemEntity.deleteSync
  pkg\dds\test\dap\integration\test_support.dart 239:13  DapTestSession.tearDown
  

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
00:06 [32m+1[0m[31m -1[0m: debug mode removes breakpoints/pause and resumes on detach[0m

00:09 [32m+1[0m[31m -2[0m: debug mode removes breakpoints/pause and resumes on detach [1m[31m[E][0m[0m

  FileSystemException: Deletion failed, path = 'C:\b\s\w\ir\x\t\dart-sdk-dap-test7ff262dd' (OS Error: The process cannot access the file because it is being used by another process.

  , errno = 32)

  dart:io/directory_impl.dart 201:7                      _Directory._deleteSync
  dart:io/file_system_entity.dart 420:7                  FileSystemEntity.deleteSync
  pkg\dds\test\dap\integration\test_support.dart 239:13  DapTestSession.tearDown
  

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
00:09 [32m+1[0m[31m -2[0m: [31mSome tests failed.[0m

stderr:
Unhandled exception:
Dummy exception to set exit code.

--- Re-run this test:
python3 tools/test.py -n unittest-asserts-release-win pkg/dds/test/dap/integration/debug_attach_test
```

Log 2:
```
--- Command "vm" (took 17.000515s):
set DART_CONFIGURATION=ReleaseX64 & set DART_SUPPRESS_WER=1 & set CHROME_PATH=C:\b\s\w\ir\cache\builder\sdk\third_party\browsers\chrome\Chrome\Application\chrome.exe & out\ReleaseX64\dart-sdk\bin\dart.exe --enable_asserts -Dtest_runner.configuration=unittest-asserts-release-win --ignore-unrecognized-flags --packages=C:\b\s\w\ir\cache\builder\sdk\.packages C:\b\s\w\ir\cache\builder\sdk\pkg\dds\test\dap\integration\debug_attach_test.dart

exit code:
255

stdout:
00:00 [32m+0[0m: debug mode can attach to a simple script using vmServiceUri[0m

00:02 [32m+1[0m: debug mode can attach to a simple script using vmServiceInfoFile[0m

00:06 [32m+2[0m: debug mode removes breakpoints/pause and resumes on detach[0m

00:10 [32m+2[0m[31m -1[0m: debug mode removes breakpoints/pause and resumes on detach [1m[31m[E][0m[0m

  Did not recieve stopped event before stream closed

  pkg\dds\test\dap\integration\test_client.dart 179:15  DapTestClient.event.<fn>
  dart:async/stream_pipe.dart 11:23                     _runUserCode
  dart:async/stream.dart 1647:9                         Stream.firstWhere.<fn>
  dart:async/zone.dart 1418:47                          _rootRun
  dart:async/zone.dart 1328:19                          _CustomZone.run
  dart:async/zone.dart 1236:7                           _CustomZone.runGuarded
  dart:async/stream_impl.dart 394:13                    _BufferingStreamSubscription._sendDone.sendDone
  dart:async/stream_impl.dart 404:7                     _BufferingStreamSubscription._sendDone
  dart:async/stream_impl.dart 610:14                    _DelayedDone.perform
  dart:async/stream_impl.dart 706:11                    _StreamImplEvents.handleNext
  dart:async/stream_impl.dart 663:7                     _PendingEvents.schedule.<fn>
  dart:async/zone.dart 1418:47                          _rootRun
  dart:async/zone.dart 1328:19                          _CustomZone.run
  dart:async/zone.dart 1236:7                           _CustomZone.runGuarded
  dart:async/zone.dart 1276:23                          _CustomZone.bindCallbackGuarded.<fn>
  dart:async/zone.dart 1426:13                          _rootRun
  dart:async/zone.dart 1328:19                          _CustomZone.run
  dart:async/zone.dart 1236:7                           _CustomZone.runGuarded
  dart:async/zone.dart 1276:23                          _CustomZone.bindCallbackGuarded.<fn>
  dart:async/schedule_microtask.dart 40:21              _microtaskLoop
  dart:async/schedule_microtask.dart 49:5               _startMicrotaskLoop
  dart:isolate-patch/isolate_patch.dart 122:13          _runPendingImmediateCallback
  dart:isolate-patch/isolate_patch.dart 193:5           _RawReceivePortImpl._handleMessage
  

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
  FileSystemException: Deletion failed, path = 'C:\b\s\w\ir\x\t\dart-sdk-dap-test3e76ba10' (OS Error: The process cannot access the file because it is being used by another process.

  , errno = 32)

  dart:io/directory_impl.dart 201:7                      _Directory._deleteSync
  dart:io/file_system_entity.dart 420:7                  FileSystemEntity.deleteSync
  pkg\dds\test\dap\integration\test_support.dart 239:13  DapTestSession.tearDown
  

Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
00:10 [32m+2[0m[31m -1[0m: [31mSome tests failed.[0m

stderr:
Unhandled exception:
Dummy exception to set exit code.

--- Re-run this test:
python3 tools/test.py -n unittest-asserts-release-win pkg/dds/test/dap/integration/debug_attach_test
```

These failures don't look related to the change.
/cc @bkonyi

**Comment by DanTup** [comment]
I believe https://dart-review.googlesource.com/c/sdk/+/231240/ will fix the first part (which is Windows failing to delete the directory for a short period after the processes exit) by retrying the delete a few times (I could reproduce this locally, and with the change it always deleted on the second attempt).

Not sure if the second one is the same (looks like "Did not recieve stopped event before stream closed" occurred first), so still looking at that.

**Comment by DanTup** [comment]
I can't come up with a reason for the other failure.. It seems like the target VM exited, without ever pausing as would be expected. Perhaps something crashed.

Do you know how often it's failed in that way? I can't see a failure on https://ci.chromium.org/p/dart/g/alt/console?limit=200 that seems to match up. If it's happening frequently, I could perhaps have the test dump the full DAP logging to get more details about what's happening and revert afterwards. If it's very infrequent that might be quite spammy to leave in the logs though.

**Comment by alexmarkov** [comment]
Out of the 5 latest `pkg-win-release` builds:

3 failures mention "Did not recieve stopped event before stream closed":

https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/19411/pkg/dds/test/dap/integration/debug_attach_test

https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/19410/pkg/dds/test/dap/integration/debug_attach_test

https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/19407/pkg/dds/test/dap/integration/debug_attach_test

2 failures only have "FileSystemException: Deletion failed":

https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/19409/pkg/dds/test/dap/integration/debug_attach_test

https://dart-ci.appspot.com/log/pkg-win-release/unittest-asserts-release-win/19408/pkg/dds/test/dap/integration/debug_attach_test

**Comment by DanTup** [comment]
Well that sounds fairly common. I'll make a change that makes the test output more verbose.

Is there somewhere I can see a list of these results myself? I usually keep my eye on https://ci.chromium.org/p/dart/g/alt/console but that seems to be green for pkg-win-release?

**Comment by alexmarkov** [comment]
Looking at the test output for the recent builds is not that easy since results feed and approval process was introduced (green build means there are no unapproved failures, it doesn't mean there are no failing tests).

I usually do the following:

1. Check current test results
https://dart-ci.firebaseapp.com/current_results/#/filter=pkg/dds/test/dap/integration/debug_attach_test
or results feed
https://dart-ci.firebaseapp.com/#test=pkg/dds/test/dap/integration/debug_attach_test
and follow the `log` link (note that results feed points to the log when the test failed for the first time).
For example
https://dart-ci.appspot.com/log/any/unittest-asserts-release-win/19412/pkg/dds/test/dap/integration/debug_attach_test
This link has a build number in it (19412) which can be replaced with another build number to get the test output in that build. The builds on CI bots are usually sequential, so if you start with the "current test results" link you can look at the recent builds by decreasing the build number (19411, 19410 etc).

2. If you need to look at the log corresponding to a particular build, you would need to find its build number. For that you can switch to the main console (https://ci.chromium.org/p/dart/g/be/console) and hover over the builds and notice links to the particular builds - these links end with the build number.
Alternatively,  click on the bot caption (top row) to open the bot page (https://ci.chromium.org/p/dart/builders/ci.sandbox/pkg-win-release) and look at the `Build #` column.

@whesse @athomas might have a better idea how to do that.

**Comment by DanTup** [comment]
@alexmarkov that's enough for me to keep an eye on them while working on fixes - thanks!

**Comment by whesse** [comment]
The process that Alex described is pretty much the best way to see the logs for recent builds - they are all uploaded to cloud storage, and the log viewer will show them.  The build number in the URL can be replaced by "latest" to get the latest build.

**Comment by DanTup** [comment]
Got it, thanks!

So I can easily find it again, a link to a failure log that includes the verbose output from the change above:

https://dart-ci.appspot.com/log/any/unittest-asserts-release-win/19427/pkg/dds/test/dap/integration/debug_attach_test

**Comment by DanTup** [comment]
I have two changes open that I believe will resolve these issues:

- add a retry to the directory delete in cleanup (I think this is a common issue on Windows, where trying to delete too quickly fails saying it's still locked)
  https://dart-review.googlesource.com/c/sdk/+/231240
- ensure the fix for a race where a PauseStart event arrives during debugger initialization also applies to attach requests (not only launch)
  https://dart-review.googlesource.com/c/sdk/+/231460

Once both are landed, I'll keep an eye on the results and if they go green, I'll revert [the change](https://github.com/dart-lang/sdk/commit/0a5da6814dab4f0013710b16987b3e7ba944233a) that made the test output more verbose.

**Comment by DanTup** [comment]
The fix above didn't work (or at least, didn't solve it completely). The problem was with how we treat paused-at-start isolates for attach. When we're attaching, we keep them paused, but subsequent isolates are auto-resumed after we've configured them.

In the case where this failed, the main isolate didn't exist when we first attached (there were 0 isolates), and then when it started it looked like a "new isolate the app started" so we auto-resumed it.

I've a change open that changes how this works. When attaching, we'll now keep all isolates paused until the user first resumes from this paused state, and then future isolates will be auto-resumed after configuring.

https://dart-review.googlesource.com/c/sdk/+/237683/
